### ISSUE TITLE
[JENKINS-8226] Add reviewers in gerrit when a build is successful

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/gerrit-trigger-plugin#558]
<p>We are using the gerrit-trigger plugin to trigger review builds.  When the build is successful we have it configured to +1 the change in gerrit.</p>

<p>This enhancement would provide for the ability to add a specific set of reviewer automatically when a build is successful.  For example, currently there are the values that allow for custom messages depending on the build results.  There could also be values for each type of build result that allows for a list of gerrit accounts to add to the review.</p>

<p>We currently have a script that uses gerrit stream-events to accomplish this, but it would be much nicer to have it integrated into hudson itself.  This seems like the logical place to do it.</p>

<p>There may be a more generic way to provide this, which would also be fine.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cygnl7">cygnl7</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-8226" target="_blank">Add reviewers in gerrit when a build is successful</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/rsandell.png" title="rsandell's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rsandell">rsandell</a>
<li><b>status</b>: Open
<li><b>priority</b>: Major
<li><b>component(s)</b>: gerrit-trigger-plugin
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 4
<li><b>watchers</b>: 8
<li><b>imported</b>: 2025-12-06
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>We are using the gerrit-trigger plugin to trigger review builds.  When the build is successful we have it configured to +1 the change in gerrit.</p>

<p>This enhancement would provide for the ability to add a specific set of reviewer automatically when a build is successful.  For example, currently there are the values that allow for custom messages depending on the build results.  There could also be values for each type of build result that allows for a list of gerrit accounts to add to the review.</p>

<p>We currently have a script that uses gerrit stream-events to accomplish this, but it would be much nicer to have it integrated into hudson itself.  This seems like the logical place to do it.</p>

<p>There may be a more generic way to provide this, which would also be fine.</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/rsandell.png" title="rsandell's avatar" /> rsandell:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/8226?focusedId=144177&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-144177">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-8226?focusedId=144177&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-144177">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This looks like a good feature. But there is no ssh command in Gerrit that adds reviewers to a change. We need to get that feature into Gerrit before we can have it in the trigger.</p>
</pre>
</details>

</ul>
</details>

<p>This looks like a good feature. But there is no ssh command in Gerrit that adds reviewers to a change. We need to get that feature into Gerrit before we can have it in the trigger.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jtbarrett:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/8226?focusedId=150549&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-150549">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-8226?focusedId=150549&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-150549">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>cygnl7, would you mind explaining how you're currently achieving this without the gerrit ssh command?</p>
</pre>
</details>

</ul>
</details>

<p>cygnl7, would you mind explaining how you're currently achieving this without the gerrit ssh command?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rasmuskvoss:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/8226?focusedId=155536&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-155536">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-8226?focusedId=155536&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-155536">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>i am also looking into this, and i believe it would be great to share the knowledge there is on this,
In version 2.2.1 there will be script 
<a href="http://egit.eclipse.org/r/Documentation/cmd-modify-reviewers.html" class="external-link" target="_blank" rel="nofollow noopener">http://egit.eclipse.org/r/Documentation/cmd-modify-reviewers.html</a>
that can handle this.
Also Shawn Pierce replied 
<a href="http://groups.google.com/group/repo-discuss/browse_thread/thread/4ce9086ef2593272/dc736c5a19ec5d56?lnk=gst&amp;q=modify+reviewers#dc736c5a19ec5d56" class="external-link" target="_blank" rel="nofollow noopener">http://groups.google.com/group/repo-discuss/browse_thread/thread/4ce9086ef2593272/dc736c5a19ec5d56?lnk=gst&amp;q=modify+reviewers#dc736c5a19ec5d56</a></p>

</pre>
</details>

</ul>
</details>

<p>i am also looking into this, and i believe it would be great to share the knowledge there is on this,<br/>
In version 2.2.1 there will be script <br/>
<a href="http://egit.eclipse.org/r/Documentation/cmd-modify-reviewers.html" class="external-link" target="_blank" rel="nofollow noopener">http://egit.eclipse.org/r/Documentation/cmd-modify-reviewers.html</a><br/>
that can handle this.<br/>
Also Shawn Pierce replied <br/>
<a href="http://groups.google.com/group/repo-discuss/browse_thread/thread/4ce9086ef2593272/dc736c5a19ec5d56?lnk=gst&q=modify+reviewers#dc736c5a19ec5d56" class="external-link" target="_blank" rel="nofollow noopener">http://groups.google.com/group/repo-discuss/browse_thread/thread/4ce9086ef2593272/dc736c5a19ec5d56?lnk=gst&q=modify+reviewers#dc736c5a19ec5d56</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>borge:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/8226?focusedId=164226&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-164226">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-8226?focusedId=164226&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-164226">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hey.</p>

<p>This feature would be really useful. I assume this is now possible by using
<a href="http://gerrit-documentation.googlecode.com/svn/Documentation/2.4.1/cmd-set-reviewers.html" class="external-link" target="_blank" rel="nofollow noopener">http://gerrit-documentation.googlecode.com/svn/Documentation/2.4.1/cmd-set-reviewers.html</a> ?
Any plans of implementing this, or could someone give pointers to where to start digging in the code to try setting it up?</p>

<p>Cheers</p>
</pre>
</details>

</ul>
</details>

<p>Hey.</p>

<p>This feature would be really useful. I assume this is now possible by using<br/>
<a href="http://gerrit-documentation.googlecode.com/svn/Documentation/2.4.1/cmd-set-reviewers.html" class="external-link" target="_blank" rel="nofollow noopener">http://gerrit-documentation.googlecode.com/svn/Documentation/2.4.1/cmd-set-reviewers.html</a> ?<br/>
Any plans of implementing this, or could someone give pointers to where to start digging in the code to try setting it up?</p>

<p>Cheers</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/rsandell.png" title="rsandell's avatar" /> rsandell:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/8226?focusedId=164590&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-164590">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-8226?focusedId=164590&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-164590">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Since I don't have time to implement it I could give some hints <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/wink.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>ToGerritRunListener is the central hub for coordinating when builds starts and completes.
when all builds for a GerritEvent are completed then NotificationFactory.queueBuildCompleted will be called that will add a BuildCompletedCommandJob to a processing queue for outgoing Gerrit commands. When it is processed it will calculate the final score(s) to send with the help of some other utility classes.</p>

<p>So you could either create a new CommandJob type that gathers the list of reviewers depending on the outcome from all builds and sends a set-reviewers command. Or add to the BuildCompletedcommandJob to also send a set-reviewers command.
I don't know which approach would be best, I think I would prefer the first alternative, but it could be a bit complex to realize compared to adding to the previous command since it has already done most of the calculations needed.</p>

<p>The configuration would be quite straight forward; the Config class holds the central configuration for how the commands themselves should look and the GerritTrigger class holds the configuration for each job.</p>

<p>Just send me a mail if you have any questions!
/B</p>
</pre>
</details>

</ul>
</details>

<p>Since I don't have time to implement it I could give some hints <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/wink.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>ToGerritRunListener is the central hub for coordinating when builds starts and completes.<br/>
when all builds for a GerritEvent are completed then NotificationFactory.queueBuildCompleted will be called that will add a BuildCompletedCommandJob to a processing queue for outgoing Gerrit commands. When it is processed it will calculate the final score(s) to send with the help of some other utility classes.</p>

<p>So you could either create a new CommandJob type that gathers the list of reviewers depending on the outcome from all builds and sends a set-reviewers command. Or add to the BuildCompletedcommandJob to also send a set-reviewers command.<br/>
I don't know which approach would be best, I think I would prefer the first alternative, but it could be a bit complex to realize compared to adding to the previous command since it has already done most of the calculations needed.</p>

<p>The configuration would be quite straight forward; the Config class holds the central configuration for how the commands themselves should look and the GerritTrigger class holds the configuration for each job.</p>

<p>Just send me a mail if you have any questions!<br/>
/B</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/rsandell.png" title="rsandell's avatar" /> rsandell:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/8226?focusedId=170590&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-170590">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-8226?focusedId=170590&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-170590">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://github.com/jenkinsci/gerrit-trigger-plugin/pull/29" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/gerrit-trigger-plugin/pull/29</a></p>
</pre>
</details>

</ul>
</details>

<p><a href="https://github.com/jenkinsci/gerrit-trigger-plugin/pull/29" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/gerrit-trigger-plugin/pull/29</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>borge:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/8226?focusedId=170608&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-170608">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-8226?focusedId=170608&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-170608">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Excellent! We were literally just about to start looking at this (yes, 4 months later <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> ). Love the separate .owners file for different areas of the app. Brilliant.</p>

<p>Any idea on when this will be merged into master and released?</p>

<p>Thanks!</p>
</pre>
</details>

</ul>
</details>

<p>Excellent! We were literally just about to start looking at this (yes, 4 months later <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> ). Love the separate .owners file for different areas of the app. Brilliant.</p>

<p>Any idea on when this will be merged into master and released?</p>

<p>Thanks!</p>
